### PR TITLE
YLP-713 Update password: hcp/caregivers must give current password

### PR DESCRIPTION
**Pull request by yvon-dblg**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,11 +2,18 @@
 
 Shoreline is the module that manages logins and user accounts.
 
-## 1.5.2 - 2021-05-03
+## Unreleased
```
Comment (franck-fourel) [inline CHANGELOG.md]: let's publish it as 1.6.0

File: user/user.go [hunk 2]
```diff
@@ -425,10 +432,12 @@ func (details *UpdateUserDetails) Validate() error {
...
-	if details.Password != nil {
-		if !IsValidPassword(*details.Password) {
-			return User_error_password_invalid
-		}
+	if details.Password != nil && !IsValidPassword(*details.Password) {
+		return User_error_password_invalid
+	}
+
+	if details.CurrentPassword != nil && !IsValidPassword(*details.CurrentPassword) {
+		return User_error_password_invalid
```
Comment (franck-fourel) [inline user/user.go]: we do not say which password is invalid. the output will be the same if current or new are invalid.

File: user/user.go [hunk 3]
```diff
@@ -452,9 +461,28 @@ func ParseUpdateUserDetails(reader io.Reader) (*UpdateUserDetails, error) {
...
+
+	if details.EmailVerified != nil {
+		details.nFields += 1
+	}
+	if len(details.Emails) > 0 {
+		details.nFields += 1
+	}
+	if details.Password != nil {
+		details.nFields += 1
+	}
```
Comment (franck-fourel) [inline user/user.go:476]: why don't we consider the currentPassword ?

File: user/api.go [hunk 4]
```diff
@@ -455,8 +458,24 @@ func (a *Api) UpdateUser(res http.ResponseWriter, req *http.Request, vars map[st
...
+		if updateUserDetails.Password != nil && !tokenData.IsServer && (originalUser.HasRole("hcp") || originalUser.HasRole("caregiver")) {
+			// Caregiver & hcp: Must provide their current password to change it
+			// Patient password change is done differently
+			// Server token: Can perform the change
+			if updateUserDetails.CurrentPassword == nil {
+				a.sendError(res, http.StatusUnauthorized, STATUS_UNAUTHORIZED, "Missing current password")
+				return
+			}
+			if !originalUser.PasswordsMatch(*updateUserDetails.CurrentPassword, a.ApiConfig.Salt) {
+				a.sendError(res, http.StatusUnauthorized, STATUS_UNAUTHORIZED, "User does not have permissions", fmt.Errorf("User '%s' passwords do not match", originalUser.Username))
```
Comment (franck-fourel) [inline user/api.go]:
```suggestion
				a.sendError(res, http.StatusUnauthorized, STATUS_PW_WRONG, "User does not have permissions", fmt.Errorf("User '%s' passwords do not match", originalUser.Username))
```

**Review by franck-fourel — CHANGES_REQUESTED** [review]
couple of questions and requests

File: user/user.go (hunk 3 above)
Comment (yvon-dblg) [reply user/user.go:476]: Because it is mandatory only if we change the password.
A change with only the current password has no action associated.

**Review by franck-fourel — APPROVED** [review]
lgtm

**Pull request merged**